lib/vssh/libssh2.c

```diff
@@ -648,6 +649,105 @@ static CURLcode ssh_check_fingerprint(struct connectdata *conn)
+#endif
+#ifdef LIBSSH2_KNOWNHOST_KEY_ECDSA_521
+  static const char * const hostkey_method_ssh_ecdsa_521
+    = "ecdsa-sha2-nistp521";
+#endif
+#ifdef LIBSSH2_KNOWNHOST_KEY_ECDSA_384
+  static const char * const hostkey_method_ssh_ecdsa_384
+    = "ecdsa-sha2-nistp284";
+#endif
+#ifdef LIBSSH2_KNOWNHOST_KEY_ECDSA_256
+  static const char * const hostkey_method_ssh_ecdsa_256
+    = "ecdsa-sha2-nistp256";
+#endif
+  static const char * const hostkey_method_ssh_rsa
+    = "ssh-rsa";
+  static const char * const hostkey_method_ssh_dss
+    = "ssh-dss";
+
+  CURLcode result = CURLE_OK;
+  const char *hostkey_method = NULL;
+  struct ssh_conn *sshc = &conn->proto.sshc;
+  struct Curl_easy *data = conn->data;
+  struct libssh2_knownhost* store = NULL;
+  bool found = false;
+
+  /*&& data->set.str[STRING_SSH_KNOWNHOSTS]
+   *<-- unnecessary because of sshc->kh? */
+  if(sshc->kh && !data->set.str[STRING_SSH_HOST_PUBLIC_KEY_MD5]) {
+    /* lets try to find our host in the known hosts file */
+    while(!libssh2_knownhost_get(sshc->kh, &store, store)) {
+      /* TODO: what about specific ports, mentioned in the name? */
+      if(strcmp(store->name, conn->host.name) == 0) {
+        infof(data, "Found host (%s) in %s\n",
+            conn->host.name, data->set.str[STRING_SSH_KNOWNHOSTS]);
+        found = true;
+        break;
+      }
+    }
+
+    if(found) {
+      /* TODO: do we need to check the host and key format here as well?*/
+      switch(store->typemask & LIBSSH2_KNOWNHOST_KEY_MASK) {
+#ifdef LIBSSH2_KNOWNHOST_KEY_ED25519
+      case LIBSSH2_KNOWNHOST_KEY_ED25519:
+        hostkey_method = hostkey_method_ssh_ed25519;
+        break;
+#endif
+#ifdef LIBSSH2_KNOWNHOST_KEY_ECDSA_521
+      case LIBSSH2_KNOWNHOST_KEY_ECDSA_521:
+        hostkey_method = hostkey_method_ssh_ecdsa_521;
+        break;
+#endif
+#ifdef LIBSSH2_KNOWNHOST_KEY_ECDSA_384
+      case LIBSSH2_KNOWNHOST_KEY_ECDSA_384:
+        hostkey_method = hostkey_method_ssh_ecdsa_384;
+        break;
+#endif
+#ifdef LIBSSH2_KNOWNHOST_KEY_ECDSA_256
+      case LIBSSH2_KNOWNHOST_KEY_ECDSA_256:
+        hostkey_method = hostkey_method_ssh_ecdsa_256;
+        break;
+#endif
+      case LIBSSH2_KNOWNHOST_KEY_SSHRSA:
+        hostkey_method = hostkey_method_ssh_rsa;
+        break;
+      case LIBSSH2_KNOWNHOST_KEY_SSHDSS:
+        hostkey_method = hostkey_method_ssh_dss;
+        break;
+      case LIBSSH2_KNOWNHOST_KEY_RSA1:
+        failf(data,
+            "Found host key type RSA1 - no clue what to do with it...\n");
+        return CURLE_SSH;
+      default:
+        failf(data, "Unknown host key type: %i\n",
+            (store->typemask & LIBSSH2_KNOWNHOST_KEY_MASK));
+        return CURLE_SSH;
+      }
+
+      infof(data, "Set \"%s\" as SSH hostkey type\n", hostkey_method);
+      /*TODO: should we add LIBSSH2_ERROR_METHOD_NOT_SUPPORTED
+       * and LIBSSH2_ERROR_INVAL to libssh2_session_error_to_CURLE()
+       * and is this even the right function at all? */
```

Comment:
Well, if you can think of a better `CURLE` code to use for those error codes than `CURLE_SSH`, then I think they should be added. Otherwise I think the default is fine. The error string is what's going to help users mostly here I think.

Reply:
I could think of CURLE_BAD_FUNCTION_ARGUMENT, but I do know if it would be useful at all. As you mentioned, the error string will help the most.